File: life-assisitant-ui-desktop/src/stores/health.ts
/**
 * Health Store（健康，记录模块的第 3 个维度）
 *
 * SYNC-FROM-BACKEND: life-assisitant-api/internal/controller/health.go
 *                    life-assisitant-api/internal/model/dto/health.go
 * SYNC-FROM-MOBILE:  life-assisitant-ui-mobile/src/stores/health.ts
 * 最后同步：2026-09-19（整体覆盖语义的「取回→合并→整体提交」封装在 saveDay）
 *
 * 缓存策略（刻意不做持久化，业务数据每次进模块拉一次）：
 *   - prediction   → 写操作后由**响应直接替换**（少一次请求，不闪旧数据）
 *   - calendar     → 按 month 缓存；写后强制刷新当月；cycles_changed 清空全部
 *   - dayDetail    → 不缓存（按需拉）
 *   - settings     → 进入模块一次；PATCH 后由响应替换
 *
 * ⚠️ 整体覆盖语义（PUT /health/days/:date）是本项目最重要的一致性约束：
 *    「没给的字段置未记录」，但 period / moods 整体不传 = 不触碰既有行。
 *    因此 saveDay 内部**必须先取回当天完整记录、合并后再整体提交**，
 *    否则快捷条（只改饮水）会抹掉当天其它指标。页面一律走 saveDay，不各自拼装。
 *
 * ⚠️ 前端不做任何预测计算：阶段、日期、区间一律渲染后端返回的 prediction。
 */
import { create } from 'zustand'
import { Toast } from '@douyinfe/semi-ui'
import { healthApi } from '@/api/health'
import type {
  CreateHealthEventReq,
  HealthCalendarResp,
  HealthDayDetail,
  HealthDayDetailResp,
  HealthEventDaySummary,
  HealthEventItem,
  HealthMoodBlock,
  HealthOverviewResp,
  HealthPeriodBlock,
  HealthSettings,
  HealthSetupReq,
  PatchHealthEventReq,
  PatchHealthSettingsReq,
  PeriodPrediction,
} from '@/api/types'

/** 本地时区当天日期 YYYY-MM-DD */
function todayDate(d: Date = new Date()): string {
  const y = d.getFullYear()
  const m = String(d.getMonth() + 1).padStart(2, '0')
  const day = String(d.getDate()).padStart(2, '0')
  return `${y}-${m}-${day}`
}

/** 一条空的当天记录（取回不到时使用，保证整体提交时其它块是「未记录」而非 undefined） */
function emptyDay(date: string): HealthDayDetail {
  return {
    date,
    water_ml: 0,
    weight_kg: null,
    bbt: null,
    sleep_hours: null,
    bowel_count: 0,
    bowel_type: 0,
    period: { flow: 0, symptoms: [], pain_level: 0, discharge: 0, intercourse: 0 },
    moods: { mood: 0, energy: 0, note: '' },
  }
}

interface HealthStore {
  // ==================== state ====================
  initialized: boolean
  today: string
  settings: HealthSettings | null
  todayLog: HealthDayDetail | null
  todayProgress: { recorded: number; enabled: number }
  cards: import('@/api/types').HealthCard[]
  monthSummary: import('@/api/types').HealthMonthSummary | null
  prediction: PeriodPrediction | null

  /** 选中日期（默认今天），记录浮层/快捷条以此为准 */
  selectedDate: string
  /** 月历缓存：month(YYYY-MM) → 响应 */
  calendarCache: Record<string, HealthCalendarResp>
  /** 当前展示的月份 */
  calendarMonth: string

  loading: boolean
  saving: boolean
  /** 概览是否已经成功拉过一次（避免引导跳转误判） */
  loaded: boolean
  /** 本次会话用户点了引导向导的「跳过」（仅内存，刷新即失效） */
  setupSkipped: boolean

  /* ===== 时间轴事件（health_events） =====
   * 身体指标一天可记 N 次、每次带时间点、**不延续**（与心情的向前延续相反）；
   * 经期块 / 心情块 / 备注仍是一天一张，走 saveDay。
   * ⚠️ 两个写入口都改同一天的展示数据，任一写成功都要刷新 overview + 当月日历。 */
  dayEvents: HealthEventItem[]
  daySummary: HealthEventDaySummary | null
  eventsLoading: boolean

  // ==================== getters ====================
  /** 是否处于「从没初始化过」→ 需要走引导（跳过态下不算需要引导） */
  needSetup: () => boolean
  /** 当前启用的指标（按 order 规整后的 key 列表） */
  enabledMetrics: () => string[]

  // ==================== actions ====================
  fetchOverview: (date?: string) => Promise<HealthOverviewResp | null>
  fetchCalendar: (month: string, force?: boolean) => Promise<HealthCalendarResp | null>
  invalidateCalendar: () => void
  fetchDay: (date: string) => Promise<HealthDayDetailResp | null>
  saveDay: (date: string, patch: Partial<HealthDayDetail>) => Promise<boolean>
  deleteDay: (date: string) => Promise<boolean>
  fetchSettings: () => Promise<HealthSettings | null>
  patchSettings: (data: PatchHealthSettingsReq) => Promise<boolean>
  setup: (data: HealthSetupReq) => Promise<boolean>
  skipSetup: () => void
  resetLocal: () => void
  fetchDayEvents: (date: string) => Promise<HealthEventItem[]>
  /** 事件写成功后统一刷新（概览 + 当月日历 + 当天时间轴） */
  refreshAfterEvent: (date: string) => Promise<void>
  addEvent: (data: CreateHealthEventReq, opts?: { silent?: boolean }) => Promise<boolean>
  editEvent: (id: string, data: Omit<PatchHealthEventReq, 'id'>) => Promise<boolean>
  removeEvent: (id: string, date: string) => Promise<boolean>
}

export const useHealthStore = create<HealthStore>((set, get) => ({
  // ==================== state ====================
  initialized: false,
  today: '',
  settings: null,
  todayLog: null,
  todayProgress: { recorded: 0, enabled: 0 },
  cards: [],
  monthSummary: null,
  prediction: null,
  selectedDate: todayDate(),
  calendarCache: {},
  calendarMonth: '',
  loading: false,
  saving: false,
  loaded: false,
  setupSkipped: false,
  dayEvents: [],
  daySummary: null,
  eventsLoading: false,

  // ==================== getters ====================
  needSetup: () => get().loaded && !get().initialized && !get().setupSkipped,
  enabledMetrics: () => get().settings?.metrics_enabled ?? [],

  // ==================== actions ====================
  async fetchOverview(date?: string) {
    set({ loading: true })
    try {
      const res = await healthApi.getOverview(date ? { date } : undefined)
      set({
        initialized: res.initialized,
        today: res.today,
        settings: res.settings,
        todayLog: res.today_log,
        todayProgress: res.today_progress,
        cards: res.cards,
        monthSummary: res.month_summary,
        prediction: res.prediction ?? null,
        loaded: true,
      })
      return res
    } catch (e) {
      // eslint-disable-next-line no-console
      console.error('[HealthStore] fetchOverview failed', e)
      set({ loaded: true })
      return null
    } finally {
      set({ loading: false })
    }
  },

  async fetchCalendar(month: string, force = false) {
    set({ calendarMonth: month })
    const cache = get().calendarCache
    if (!force && cache[month]) return cache[month]
    try {
      const res = await healthApi.getCalendar({ month })
      set({ calendarCache: { ...get().calendarCache, [month]: res } })
      // 月历响应里也带 prediction（切月后概览一并刷新）
      if (res.prediction) set({ prediction: res.prediction })
      return res
    } catch (e) {
      // eslint-disable-next-line no-console
      console.error('[HealthStore] fetchCalendar failed', e)
      return null
    }
  },

  invalidateCalendar() {
    set({ calendarCache: {} })
  },

  async fetchDay(date: string) {
    try {
      return await healthApi.getDay(date)
    } catch (e) {
      // eslint-disable-next-line no-console
      console.error('[HealthStore] fetchDay failed', e)
      return null
    }
  },

  /**
   * 保存某天记录（整体覆盖语义的核心封装）。
   * 流程：取回当天完整记录 → 合并 patch → 整体提交 → 刷新 overview + 当月日历。
   *
   * ⚠️ patch 只传要改的字段即可；period / moods 子块**缺省不传**时沿用既有行，
   *    传了则整块覆盖（记录浮层走整块覆盖，快捷条只传标量字段不碰子块）。
   *
   * @returns 是否成功
   */
  async saveDay(date: string, patch: Partial<HealthDayDetail>) {
    set({ saving: true })
    try {
      // 1) 取回当天完整记录（这是整体覆盖语义不抹数据的前提）
      const res = await healthApi.getDay(date)
      const base = res?.day ?? emptyDay(date)
      // 2) 合并：标量字段覆盖，period/moods 子块整体覆盖或沿用
      const merged: HealthDayDetail = {
        ...base,
        ...patch,
        period: (patch.period as HealthPeriodBlock | undefined) ?? base.period,
        moods: (patch.moods as HealthMoodBlock | undefined) ?? base.moods,
      }
      // 3) 整体提交
      const saveRes = await healthApi.upsertDay(date, merged)

      if (saveRes.prediction) set({ prediction: saveRes.prediction })
      if (saveRes.cycles_changed) get().invalidateCalendar()
      if (date === get().today) set({ todayLog: saveRes.day })

      // 4) 成功后刷新 overview + 当月日历（保证概览卡/完成度/月历同步）
      await get().fetchOverview()
      if (get().calendarMonth) await get().fetchCalendar(get().calendarMonth, true)

      Toast.success('已记录')
      return true
    } catch (e) {
      // 失败：清掉陈旧数据，避免展示与后端不一致的本地值
      if (date === get().today) set({ todayLog: null })
      // eslint-disable-next-line no-console
      console.error('[HealthStore] saveDay failed', e)
      Toast.error('保存失败，请重试')
      return false
    } finally {
      set({ saving: false })
    }
  },

  async deleteDay(date: string) {
    set({ saving: true })
    try {
      await healthApi.deleteDay(date)
      get().invalidateCalendar()
      // 删除接口无响应体，只能重拉
      const res = await healthApi.getOverview()
      set({
        initialized: res.initialized,
        today: res.today,
        settings: res.settings,
        todayLog: res.today_log,
        todayProgress: res.today_progress,
        cards: res.cards,
        monthSummary: res.month_summary,
        prediction: res.prediction ?? null,
      })
      Toast.success('已删除')
      return true
    } catch (e) {
      // eslint-disable-next-line no-console
      console.error('[HealthStore] deleteDay failed', e)
      Toast.error('删除失败，请重试')
      return false
    } finally {
      set({ saving: false })
    }
  },

  async fetchSettings() {
    try {
      const s = await healthApi.getSettings()
      set({ settings: s })
      return s
    } catch (e) {
      // eslint-disable-next-line no-console
      console.error('[HealthStore] fetchSettings failed', e)
      return null
    }
  },

  async patchSettings(data: PatchHealthSettingsReq) {
    set({ saving: true })
    try {
      const s = await healthApi.patchSettings(data)
      set({ settings: s })
      // 指标/目标变化会影响概览卡与完成度，顺手刷新
      await get().fetchOverview()
      Toast.success('已保存')
      return true
    } catch (e) {
      // eslint-disable-next-line no-console
      console.error('[HealthStore] patchSettings failed', e)
      Toast.error('保存失败，请重试')
      return false
    } finally {
      set({ saving: false })
    }
  },

  async setup(data: HealthSetupReq) {
    set({ saving: true })
    try {
      const res = await healthApi.setup(data)
      set({ settings: res.settings })
      if (res.prediction) set({ prediction: res.prediction })
      set({ initialized: true, setupSkipped: false })
      get().invalidateCalendar()
      await get().fetchOverview()
      Toast.success('健康模块已开启')
      return true
    } catch (e) {
      // eslint-disable-next-line no-console
      console.error('[HealthStore] setup failed', e)
      Toast.error('保存失败，请重试')
      return false
    } finally {
      set({ saving: false })
    }
  },

  skipSetup() {
    set({ setupSkipped: true })
  },

  /* ==================== events 读写 ==================== */

  async fetchDayEvents(date: string) {
    set({ eventsLoading: true })
    try {
      const res = await healthApi.listEvents({ date })
      set({ dayEvents: res.items ?? [], daySummary: res.summary ?? null })
      return get().dayEvents
    } catch (e) {
      // eslint-disable-next-line no-console
      console.error('[HealthStore] fetchDayEvents failed', e)
      set({ dayEvents: [], daySummary: null })
      return []
    } finally {
      set({ eventsLoading: false })
    }
  },

  /** 事件写成功后统一刷新（概览卡 + 当月日历 + 当天时间轴） */
  async refreshAfterEvent(date: string) {
    await get().fetchOverview()
    const m = get().calendarMonth
    if (m) await get().fetchCalendar(m, true)
    await get().fetchDayEvents(date)
  },

  /**
   * 追加一次记录。
   * ⚠️ time 不传 = 服务端当前时刻，前端**不要**自己算（客户端时钟可能不准 / 跨时区）。
   */
  async addEvent(data: CreateHealthEventReq, opts?: { silent?: boolean }) {
    set({ saving: true })
    try {
      const res = await healthApi.createEvent(data)
      if (res.day !== undefined && data.date === get().today) set({ todayLog: res.day })
      await get().refreshAfterEvent(data.date)
      // ⚠️ 高频操作（如概览卡「＋ 喝一杯」）不弹成功 toast，就地反馈即可
      if (!opts?.silent) Toast.success('已记录')
      return true
    } catch (e) {
      // eslint-disable-next-line no-console
      console.error('[HealthStore] addEvent failed', e)
      Toast.error('保存失败，请重试')
      return false
    } finally {
      set({ saving: false })
    }
  },

  async editEvent(id: string, data: Omit<PatchHealthEventReq, 'id'>) {
    set({ saving: true })
    try {
      const res = await healthApi.patchEvent(id, data)
      if (res.day !== undefined && res.item?.date === get().today) set({ todayLog: res.day })
      await get().refreshAfterEvent(res.item?.date ?? get().selectedDate)
      Toast.success('已更新')
      return true
    } catch (e) {
      // eslint-disable-next-line no-console
      console.error('[HealthStore] editEvent failed', e)
      Toast.error('保存失败，请重试')
      return false
    } finally {
      set({ saving: false })
    }
  },

  async removeEvent(id: string, date: string) {
    set({ saving: true })
    try {
      await healthApi.deleteEvent(id)
      await get().refreshAfterEvent(date)
      Toast.success('已删除')
      return true
    } catch (e) {
      // eslint-disable-next-line no-console
      console.error('[HealthStore] removeEvent failed', e)
      Toast.error('删除失败，请重试')
      return false
    } finally {
      set({ saving: false })
    }
  },

  resetLocal() {
    set({
      initialized: false,
      today: '',
      settings: null,
      todayLog: null,
      todayProgress: { recorded: 0, enabled: 0 },
      cards: [],
      monthSummary: null,
      prediction: null,
      selectedDate: todayDate(),
      calendarCache: {},
      calendarMonth: '',
      loading: false,
      saving: false,
      loaded: false,
      setupSkipped: false,
      dayEvents: [],
      daySummary: null,
      eventsLoading: false,
    })
  },
}))

export default useHealthStore

File: life-assisitant-api/internal/model/dto/health.go
// Package dto 健康模块传输对象（md/spec-20260919-v1/07 §1）
//
// 设计约束：
//
//  1. 一天的记录由**三张表**拼成（07 §4 事务边界）：
//
//     health_days  饮水 / 体重 / 体温 / 睡眠 / 排便
//     period_days  经量 / 症状 / 疼痛 / 分泌物 / 性生活
//     mood_logs    心情 / 精力 / 备注（**按小时**，260919 改造）
//
//  2. 写接口是「整体覆盖」：请求体里没给的字段一律置为「未记录」。
//     唯一的例外是 period / moods 两个**子块**：子块整体缺省 = 不触碰既有行
//     —— 记体重时不该把当天已经填过的经期事实抹掉。
//
//  3. 心情按小时存，所以 moods 块落库时要选一个目标小时：
//     记今天 = 服务端当前小时；补记历史某天 = model.MoodDiaryHour（12，一天一条槽位）。
package dto

import "strings"

// ============================================================================
// 通用块
// ============================================================================

// HealthPeriodBlock 经期事实块（period_days）
type HealthPeriodBlock struct {
	// Flow 经量 0–4（0 = 未记录 / 无；1 = 点滴；2/3/4 = 少/中/多）
	Flow int `json:"flow"`
	// Symptoms 症状 key 数组，命中经期词库；未知 key 服务端静默丢弃
	Symptoms []string `json:"symptoms"`
	// PainLevel 疼痛程度 0–3
	PainLevel int `json:"pain_level"`
	// Discharge 分泌物 0–3
	Discharge int `json:"discharge"`
	// Intercourse 性生活 0/1
	Intercourse int `json:"intercourse"`
}

// HealthMoodBlock 心情块（mood_logs 的某一小时）
//
// ⚠️ mood_logs 是按小时的（UNIQUE(user_id,date,hour)），本块只是「一天一张表单」
// 的视图。读取时取：记今天 → 当前小时（含向前延续）；补记历史 → 日记槽位 hour=12。
type HealthMoodBlock struct {
	Mood   int    `json:"mood"`   // 1–5，0 = 未记录
	Energy int    `json:"energy"` // 1–3，0 = 未记录
	Note   string `json:"note"`   // ≤50 字；不跨小时延续
}

// HealthDayDetail 一天的完整健康记录（三表拼装结果）
type HealthDayDetail struct {
	Date string `json:"date"`

	// ↓ health_days
	WaterML    int      `json:"water_ml"`
	WeightKG   *float64 `json:"weight_kg"`
	BBT        *float64 `json:"bbt"`
	SleepHours *float64 `json:"sleep_hours"`
	BowelCount int      `json:"bowel_count"`
	BowelType  int      `json:"bowel_type"` // 0–4（见 model.BowelType*）

	// ↓ period_days（未启用经期时仍返回零值块，避免前端到处判空）
	Period *HealthPeriodBlock `json:"period"`
	// ↓ mood_logs
	Moods *HealthMoodBlock `json:"moods"`
	// ↓ health_events（时间轴明细，按 time 升序）
	//
	// ⚠️ 2026-09-19 起身体指标改为**多次时间轴记录**，这里带上当天全部事件，
	// 浮层打开就能直接画时间轴，不必再调一次 /health/events。
	// nil 与空数组都表示「这天没有事件」，前端用 `events?.length ?? 0` 判断即可。
	Events []HealthEventItem `json:"events,omitempty"`
}

// IsEmpty 三个字段全空（用于判断「这一天到底有没有心情记录」）
//
// ⚠️ 注意区分：Mood=0 只代表「这一小时没单独记心情」，整体为空才代表没记录。
func (m *HealthMoodBlock) IsEmpty() bool {
	return m == nil || (m.Mood == 0 && m.Energy == 0 && strings.TrimSpace(m.Note) == "")
}

// HealthLastValues 供「沿用上次」chip（07 §1.4）
//
// ⚠️ 只装 health_days 的两项：体重 / 体温在 2026-09-19 已从 period_days 迁走，
// 所以叫 health_last_values 而不是旧经期模块的 last_values。
type HealthLastValues struct {
	BBT      *float64 `json:"bbt"`
	WeightKG *float64 `json:"weight_kg"`
}

// ============================================================================
// 1. GET /health/overview
// ============================================================================

// HealthOverviewReq 概览查询
type HealthOverviewReq struct {
	Date string `json:"date" dc:"以哪一天为「今天」计算，默认服务端当天"`
}

// HealthSettingsResp 健康设置（每用户一行，懒创建 → GET 永不 404）
type HealthSettingsResp struct {
	MetricsEnabled []string `json:"metrics_enabled"`
	WaterGoalML    int      `json:"water_goal_ml"`
	// WaterStepML 「一杯」的容量 ml（快捷加水步进）。老行回落 200，见 model.NormalizeWaterStep
	WaterStepML  int      `json:"water_step_ml"`
	WeightGoalKG *float64 `json:"weight_goal_kg,omitempty"`
	// SetupDoneAt RFC3339；未引导时为空串
	SetupDoneAt string `json:"setup_done_at,omitempty"`
}

// HealthTodayProgress 今日完成度（分母只数**启用且可判定**的指标）
type HealthTodayProgress struct {
	Recorded int `json:"recorded"`
	Enabled  int `json:"enabled"`
}

// HealthCard 概览卡的指标行
//
// ⚠️ Tone 只取 normal / neutral —— 刻意不提供 good / bad，
// 防止前端拿去做价值判断给身体上色（07 §1.1）。
//
// ⚠️ 值拼装按 02 §14.2 五型分支（目标型 `1200 / 1500 ml`、次/日型 `55.2 kg`、
// 事件型 `2 次`、周期型 `周期第 26 天`、小时型取最近一次、未记录一律 `—`）。
// Count / State / Latest / Icon 为 02 §14.6 新增字段，供分类卡渲染「x 次」、色条、副值等。
type HealthCard struct {
	Key       string   `json:"key"`
	Label     string   `json:"label"`
	Value     string   `json:"value"` // 主值（未记录 = "—"）
	Progress  *float64 `json:"progress,omitempty"`
	DeltaText string   `json:"delta_text,omitempty"` // 体重等：较上周变化
	// Count 今日记录条数（来自 health_events 的 COUNT，周期/经期类为当日事实数）
	Count int `json:"count,omitempty"`
	// State 记录状态：recorded（已记录）/ empty（未记录，显示 —）/ n/a（周期型未到日子）
	State string `json:"state,omitempty"`
	// Latest 小时型指标（心情/精力）最近一次的时刻，如 "14:00"
	Latest string `json:"latest,omitempty"`
	// Icon 格子左侧图标（Lucide name），可选
	Icon string `json:"icon,omitempty"`
	Tone string `json:"tone"`
}

// 卡片色调
const (
	CardToneNormal  = "normal"
	CardToneNeutral = "neutral"
)

// HealthMonthSummary 月份指标摘要条（月历上方的那一行）
//
// 字段**按启用指标动态填充**：未启用的指标对应字段为零值，由调用方决定是否展示。
type HealthMonthSummary struct {
	Month         string   `json:"month"`
	RecordedDays  int      `json:"recorded_days"`
	WaterGoalDays int      `json:"water_goal_days"`
	WeightAvg     *float64 `json:"weight_avg,omitempty"`
	PeriodDays    int      `json:"period_days"`
}

// HealthOverviewResp 概览响应（一次拉齐，避免前端连打 4 个接口）
type HealthOverviewResp struct {
	Initialized   bool                 `json:"initialized"` // false → 前端进引导页
	Today         string               `json:"today"`
	Settings      *HealthSettingsResp  `json:"settings"`
	TodayLog      *HealthDayDetail     `json:"today_log"` // 无记录 = null
	TodayProgress *HealthTodayProgress `json:"today_progress"`
	// Cards 顺序与指标注册表一致；未启用的指标不出现
	Cards        []HealthCard        `json:"cards"`
	MonthSummary *HealthMonthSummary `json:"month_summary"`
	// Prediction 经期预测；未启用经期时为 null（结构见 dto.PeriodPrediction）
	Prediction *PeriodPrediction `json:"prediction,omitempty"`
}

// ============================================================================
// 2. GET /health/calendar
// ============================================================================

// HealthCalendarReq 月历查询
type HealthCalendarReq struct {
	Month string `json:"month" v:"required|length:7,7" dc:"YYYY-MM"`
}

// HealthCalendarDay 有标记的一天（**无标记的日期不返回**，否则响应体积翻 3 倍）
type HealthCalendarDay struct {
	Date  string   `json:"date"`
	Marks []string `json:"marks"`
}

// HealthCalendarResp 月历响应
type HealthCalendarResp struct {
	Month string              `json:"month"`
	Days  []HealthCalendarDay `json:"days"`
}

// ============================================================================
// 3. GET /health/days
// ============================================================================

// ListHealthDaysReq 区间查询（跨度上限 366 天，由 service 校验）
type ListHealthDaysReq struct {
	Start string `json:"start" dc:"YYYY-MM-DD"`
	End   string `json:"end" dc:"YYYY-MM-DD"`
}

// ListHealthDaysResp 区间记录
type ListHealthDaysResp struct {
	Items []HealthDayDetail `json:"items"`
	Total int               `json:"total"`
}

// ============================================================================
// 4. GET /health/days/:date
// ============================================================================

// HealthDayDetailResp 单日详情（打开浮层时用）
//
// ⚠️ 无记录时 Day = null 且**接口仍返回 200**（浮层要打开一张空表），不是 404。
type HealthDayDetailResp struct {
	Day *HealthDayDetail `json:"day"`
	// RecentSymptoms 近 90 天出现次数降序取前 6；从未记录过 = 空数组（前端据此隐藏分组）
	RecentSymptoms   []string          `json:"recent_symptoms"`
	HealthLastValues *HealthLastValues `json:"health_last_values"`
}

// ============================================================================
// 5. PUT /health/days/:date
// ============================================================================

// UpsertHealthDayReq 写某日健康记录（整体覆盖）
//
// 数值字段用**零值**表示「未记录」，所以不需要指针；
// 只有允许显式 NULL 的三项（体重 / 体温 / 睡眠）用指针区分「没给」与「给 null」。
// ⚠️ 2026-09-19 起变更（老大复核 spec 时提出）：
//
//	water_ml / weight_kg / bbt / sleep_hours / bowel_count / bowel_type 六项**已从本请求移除**。
//	它们的真源变成 health_events（一天多次、不延续），日汇总由后端在写 events 时
//	同一事务内重算写回 health_days —— **这里是只读的**，传了也无效。
//
//	要改身体指标请调 POST/PATCH/DELETE /health/events。
//	留着这些字段只会造成「两个写入口、语义冲突」（整体覆盖 vs 追加一条）。
type UpsertHealthDayReq struct {
	Date string `json:"date"`

	// Period 经期块；**nil = 不触碰既有 period_days 行**
	Period *HealthPeriodBlock `json:"period"`
	// Moods 心情块；**nil = 不触碰既有 mood_logs 行**
	Moods *HealthMoodBlock `json:"moods"`
}

// UpsertHealthDayResp 写响应
type UpsertHealthDayResp struct {
	Day *HealthDayDetail `json:"day"`
	// CyclesChanged period 块影响了经期划分 → true（前端可据此提示「周期已更新」）
	CyclesChanged bool `json:"cycles_changed"`
	// Prediction 重算后的预测，随响应返回，前端不必二次请求
	Prediction *PeriodPrediction `json:"prediction,omitempty"`
}

// ============================================================================
// 7. GET / PATCH /health/settings
// ============================================================================

// PatchHealthSettingsReq 只传要改的字段
//
// ⚠️ MetricsEnabled 为 nil 表示不动；传空数组 [] 表示**用户主动全关**（合法状态）。
// 这个 nil / [] 的区别很重要，别用 len()==0 一并处理。
type PatchHealthSettingsReq struct {
	MetricsEnabled []string `json:"metrics_enabled"`
	WaterGoalML    *int     `json:"water_goal_ml"`
	// WaterStepML 为 nil 表示不动；传 0 或越界 → 归一到 [50,1000] 的 50 倍数
	WaterStepML  *int     `json:"water_step_ml"`
	WeightGoalKG *float64 `json:"weight_goal_kg"`
}

// ============================================================================
// 8. POST /health/setup
// ============================================================================

// HealthPeriodInit 引导里的经期初始化（仅当勾选 period 时携带）
type HealthPeriodInit struct {
	LastPeriodStart string `json:"last_period_start"`
	AvgPeriodLength int    `json:"avg_period_length"`
	AvgCycleLength  int    `json:"avg_cycle_length"`
	Goal            int    `json:"goal"`
}

// HealthSetupReq 首次引导一次性提交
//
// 把「勾选指标 + 饮水目标 + 经期初始化」压成一次请求：
// 分三次提交的话，填到一半断网会留下半个状态（勾选了指标但没初始化经期）。
type HealthSetupReq struct {
	MetricsEnabled []string `json:"metrics_enabled"`
	WaterGoalML    int      `json:"water_goal_ml"`
	// WaterStepML 「一杯」的容量 ml；0 或越界 → 归一到 200
	WaterStepML int               `json:"water_step_ml"`
	PeriodInit  *HealthPeriodInit `json:"period_init"`
}

// HealthSetupResp 引导响应
type HealthSetupResp struct {
	Settings   *HealthSettingsResp `json:"settings"`
	Prediction *PeriodPrediction   `json:"prediction,omitempty"`
	// CreatedDays 批量生成的初始 period_days 行数
	CreatedDays int `json:"created_days"`
}
